Public/Sdk/Public/Deployment/operations.dsc

```diff
@@ -240,7 +240,7 @@ function flattenStaticDirectory(staticDirectory: StaticDirectory, targetFolder:
                 // TODO: Validate if there is a flattenedFile already under this OpaqueDirectory. To implement this we'll need IsWithin on RelativePath
                 return {
                     flattenedFiles: result.flattenedFiles,
-                    flattenedOpaques: result.flattenedOpaques.add(targetFolder, <OpaqueDirectory>staticDirectory),
+                    flattenedOpaques: result.flattenedOpaques.add(targetFolder, [<OpaqueDirectory>staticDirectory, r`.`]),

```

Comment:
> r`.` [](start = 114, length = 5)

Ok, so this is an example of the comment I made on definition.dsc. In this case we don't really need a designated subdir.
My suggestion is to define the flattened opaque member as something like

Map<RelativePath, OpaqueDirectory | OpaqueDirectoryWithDesignatedSubdir]>

Reply:
Or maybe OpaqueSubDirectory, which you already defined, is good enough instead of this 'OpaqueDirectoryWithDesignatedSubdir' that I was suggesting?

---
In reply to: [330193191](https://github.com/microsoft/BuildXL/pull/966#discussion_r330193191) [](ancestors = 330193191)

Reply:
How can you then easily distinguish between `OpaqueDirectory` and `OpaqueDirectoryWithDesignatedSubdir`?   I'm not sure if `typeof` returns something other than `object` when invoked against an OpaqueDirectory; if it does, then I like this proposal

Reply:
Now I think the type should just be `Map<RelativePath, OpaqueSubDirectory>`. 

I kind of like using a single type (like `OpaqueSubDirectory`) with both "parent opaque directory" and "relative path" properties than using a union type (mostly because it avoids any kind of checking/pattern matching to determine if the object is of type OpaqueDirectory or OpaqueSubDirectory)

Reply:
yes, I think typeof won't work. But you can use a known field to distinguish them (in a function where you would use a type predicate to narrow down the result as well)

---
In reply to: [330194376](https://github.com/microsoft/BuildXL/pull/966#discussion_r330194376) [](ancestors = 330194376)

Reply:
Alright, let's give union type a try here

Reply:
What do you think about using only a non-union type (like `OpaqueSubDirectory`) where the `subDirectory` property is optional?

In that case, one would here write:
```typescript
                    flattenedOpaques: result.flattenedOpaques.add(targetFolder, {opaque: <OpaqueDirectory>staticDirectory}),
 
```

Reply:
yes, that also works for me

---
In reply to: [330202985](https://github.com/microsoft/BuildXL/pull/966#discussion_r330202985) [](ancestors = 330202985)